メモアプリの開発

takao0206 opened the pull request:
主要な機能の開発が完了しましたので、一度pushしました。

takao0206 committed:
主要な機能の開発完了

takao0206 committed:
README.mdのセットアップの手順を変更。

takao0206 commented:
メモアプリの開発、確認が完了しました。
ご都合の良いときにご確認いただければ幸いです。

yoshitsugu commented on Gemfile:
```diff
@@ -0,0 +1,11 @@
+
+source 'https://rubygems.org'
+
+gem 'erb_lint', '~> 0.6.0', require: false
```
rubocopやerb_lintなど、開発環境でしか使わないものは、bundlerのgroup機能を使って、 `group :development do` というブロックの中にいれるとよいです。
https://bundler.io/guides/groups.html

takao0206 replied:
Gemfileの書き方のアドバイス、ありがとうございます。
本番でも使うgem、開発でしか使わないgemに分けたいと思います。

yoshitsugu commented on public/myscripts.js:
```diff
@@ -0,0 +1,33 @@
+function confirmDelete() {
+  const confirm = window.confirm("削除を実行しますか？");
```
confirmationのモーダルを実装できていてよいですね :+1: 
スクリプト名が `myscript` だとちょっと意味がわからないので、たとえば `form_helper` など役割がわかるようなものにしたほうがいいかなと思いました。

takao0206 replied:
ファイル名のアドバイス、ありがとうございます。
修正させていただきます。

yoshitsugu commented on views/new.erb:
```diff
@@ -0,0 +1,19 @@
+
+  <div class="form-group">
+    <label for="title">タイトル：</label>
+    <input type="text" id="title" name="title" class="form-control" value="<%= h(session[:title]) %>">
```
なぜ `session` が使われているのでしょうか？

takao0206 replied:
こちらの`session`につきまして、修正漏れになります。申し訳ございませんでした。

背景：
- 実は、JavaScriptでconfirmationのモーダルを作る前に、Rubyでモーダルのような入力値をチェックする機能を作ろうと思っていました。
- それは「空欄だったら、リダイレクトして、同じ画面に戻る」ような機能でしたが、リダイレクト後、`title`や`content`に入力した文字が消えてしまいました。
- そこで、`session`をつかって、`title`や`content`に入力した文字を保持するようにしていました。
- しかし、最終的には、JavaScriptでモーダル機能をつくりましたので、sessionは不要になりました。
- 提出前に、`session`に関するコードを削除したつもりでしたが、一部残ってしまっておりました。

yoshitsugu commented on memo_app.rb:
```diff
@@ -0,0 +1,100 @@
+  end
+
+  memos = JSON.parse(File.read(MEMOS_PATH))
+  erb :index, locals: { memos: }
```
これでもよいのですが、基本的にroute handlerからtemplateへの受け渡しはインスタンス変数を使うことが多いです。
https://sinatrarb.com/intro.html#accessing-variables-in-templates

> This is typically used when rendering templates as partials from within other templates.

と書いている通り、localsはpartialと呼ばれる部分的なviewに使うことが多いです。

takao0206 replied:
変数のアドバイス、ありがとうございます。
積極的に変数の使う範囲を絞りたい（特定のテンプレートだけに変数を渡したい）以外は、インスタンス変数を用いるようにしたいと思います。

yoshitsugu commented on memo_app.rb:
```diff
@@ -0,0 +1,100 @@
+end
+
+post '/memos' do
+  id = SecureRandom.uuid
+  title = params[:title]
+  content = params[:content]
+
+  memo = { id:, title:, content: }
```
一度ローカル変数にいれる意味があまりないかなと思いました。

```suggestion
  memo =
    {
      id: SecureRandom.uuid,
      title: params[:title],
      content: params[:content],
    }
```

takao0206 replied:
アドバイス、ありがとうございます。
おっしゃるとおり、ローカル変数にする意味（例えば、複数回使うなど）がありませんので、ご提案のように修正したいと思います。

yoshitsugu commented on memo_app.rb:
```diff
@@ -0,0 +1,100 @@
+  memos << memo
+
+  File.open(MEMOS_PATH, 'w') do |file|
+    file.write(JSON.pretty_generate(memos))
```
このようにデータを保存する場合、ファイルを直接見ることはあまりない、かつ見る場合も見るときに加工すればいいので、あまり保存時 pretty_generate を使う意味は薄いかなと思います。むしろ余計な空白などが増えて、ファイルサイズが肥大してしまいそうです。

takao0206 replied:
おっしゃるとおり、JSONファイルを直接見る機会はほとんどありませんでした。
`JSON.pretty_generate()`ではなく、`JSON.generate()`に変更したいと思います。

yoshitsugu commented on memo_app.rb:
```diff
@@ -0,0 +1,100 @@
+end
+
+get '/memos/:id/edit' do
+  memos = JSON.parse(File.read(MEMOS_PATH))
```
ファイルからの読み込み部分とファイルへの書き込み部分はメソッド化すると共通化できてよいと思います。

takao0206 replied:
おっしゃるとおり、 JSON.parse(File.read(MEMOS_PATH))を繰り返し使っておりますので、メソッド化したいと思います。

yoshitsugu commented on memo_app.rb:
```diff
@@ -0,0 +1,100 @@
+  memos = JSON.parse(File.read(MEMOS_PATH))
+
+  memo = memos.find { |m| m['id'] == id }
+  memo['title'] = new_title if memo
+  memo['content'] = new_content if memo
```
```suggestion
  if memo
    memo['title'] = new_title
    memo['content'] = new_content
  end
```

もしくは、ない場合そのままリダイレクトするとよいと思います。

takao0206 replied:
アドバイス、ありがとうございます。
今回は、`if memo ... end` の中に、コードを入れるように変更したいと思います。

yoshitsugu commented on public/myscripts.js:
```diff
@@ -0,0 +1,33 @@
+    errorMessage.style.display = "none";
+    errorMessage.innerText = "";
+
+    if (!memoTitle || !memoContent) {
```
バリデーションも考慮していますね :+1: 
簡易的なものであればHTMLのrequired属性でも十分かなと思いました。
https://developer.mozilla.org/ja/docs/Web/HTML/Attributes/required

takao0206 replied:
HTMLのrequired属性の情報、ありがとうございます。
確認して、一度required属性でバリエーションの機能を作ってみたいと思います。

yoshitsugu commented on public/myscripts.js:
```diff
@@ -0,0 +1,33 @@
+window.onload = function () {
+  const inputElements = document.getElementById('title');
+  if (inputElements) {
+    inputElements.focus();
```
こちらもHTML属性の `autofocus` で十分かなと思います。
https://developer.mozilla.org/ja/docs/Web/HTML/Global_attributes/autofocus

takao0206 replied:
`autofocus`機能のご紹介、ありがとうございます。
MDN Web Docsでの調査が不足していたようです。
フロントエンドの開発の時は、こちらのサイトも確認しながら開発するようにしたいと思います。

yoshitsugu reviewed (commented):
コメントしました！

takao0206 commented:
詳細までチェックいただき、ありがとうございました。
修正後、いつものように、ご連絡させていただきたいと思います。
少しお待ちいただければ幸いです。

takao0206 committed:
修正: Gemfileの書き方として、dev用のgemを分けた。

takao0206 committed:
修正: new.erbの中のsessionの誤記をparamsに変更。

takao0206 committed:
変更: ファイルの読み込み、書き込みについて、繰り返している部分をメソッド化した。

takao0206 committed:
変更: 生成ファイル内で無用の空白や改行を作らないため、`pretty_generate`を`generate`に変更。

takao0206 committed:
変更: 不要なローカル変数を削除。

takao0206 committed:
変更: if文の書き方を見直し。

takao0206 committed:
変更: ファイルの検索について、繰り返している部分をメソッド化した。

takao0206 committed:
変更: route handlerからtemplateに渡す変数をインスタンス化した。

takao0206 committed:
変更: HTMLグローバル属性`autofocus`を追加。JavaScriptによる同機能を削除。

takao0206 committed:
変更: HTML`required`属性を追加。JavaScriptによる同機能を削除し、JavaScriptのファイル名を機能が分かるものに変更。

takao0206 commented:
ご指摘いただきました点、修正致しました。
ご都合の良いときにご確認いただければ幸いです。

yoshitsugu commented on memo_app.rb:
```diff
@@ -0,0 +1,105 @@
+end
+
+not_found do
+  status 404
```
この `not_found` メソッドは基本的に404を返すか、 `Sinatra::NotFound` が発生したときに実行されるものなので、ここで `status 404` は冗長に思いました。
https://sinatrarb.com/intro.html#not-found

takao0206 replied:
コメントいただき、ありがとうございました。
`status 404`を削除した後、存在しないURL にアクセスしたところ、
「"GET /not_exist HTTP/1.1" 404 」が返ってくることを確認できました。
`status 404`を削除したいと思います。

yoshitsugu commented on memo_app.rb:
```diff
@@ -0,0 +1,105 @@
+  if @memo
+    erb :memo
+  else
+    status 404
+    erb :not_found
```
`not_found` メソッドの説明の通り、数値の404を返すだけで `not_found` メソッドが実行されるので、ここであらためて `erb :not_found` を指定する必要はないかなと思います。

```suggestion
    404
```

takao0206 replied:
以下のコードで動作確認を行いました。
おっしゃるとおり、ステータスも404を返して、Webブラウザにも「Not Found」と表示されました。
以下のコードで、十分に機能を果たしていると思います。
```
not_found do
end
```

yoshitsugu replied:
僕が書いたのは、ここでは `404` を返すとsinataraの仕組みで `not_found` が実行されるので、 `erb` の設定などはそちらにまとめられますよ、ということでした。このままだとここで `erb` も設定しているので、あまり `not_found` メソッドの意味がないです。

```ruby
get '/memos/:id' do
  memos = load_memos
  @memo = find_memo(memos, params[:id])

  if @memo
    erb :memo
  else
    404
  end
end

# 404のときはこちらが読み込まれる
not_found do
  erb :not_found
end
```

yoshitsugu commented on memo_app.rb:
```diff
@@ -0,0 +1,105 @@
+
+get '/memos/:id/edit' do
+  memos = load_memos
+  @memo = find_memo(memos, params[:id])
```
ここでも `@memo` が見つからなかった場合も考慮するとよいかなと思います。

takao0206 replied:
コメントいただき、ありがとうございます。
`/memos/:id/edit`で、存在しない`:id`を指定した時、Sinatraのエラー画面が表示されました。
Sinatraのエラー画面が出るのは良くないと思いますので、ここにも以下のコードを追記したいと思います。
```
  if @memo
    erb :memo
  else
    status 404
    erb :not_found
  end
  ```

yoshitsugu commented on public/form_helper.js:
```diff
@@ -0,0 +1,4 @@
+function confirmDelete() {
+  const confirm = window.confirm("削除を実行しますか？");
+  return confirm
+}
```
この程度であれば簡易的にHTML側で `onClick="return confirm('削除を実行しますか？');"` としてもよいと思います。このままでもOKなので、おまかせします。

takao0206 replied:
HTMLに`onclick="return confirm('削除を実行しますか？');"`を埋め込みましたところ、JavaScriptと同様に機能を実現できました。
この方がシンプルですので、変更（そして外部ファイルのJavaScriptを削除）したいと思います。
今後、フロントエンド（JavaScriptやReact）のプラクティスもあると思いますが、まずはHTMLでできることを再勉強したいと思います。

yoshitsugu reviewed (commented):
コメントしました！

takao0206 commented:
ご確認いただき、ありがとうございました。
今回も大変勉強になりました。
プログラムの修正後、再度ご連絡させていただきたいと思います。

takao0206 committed:
下記の修正を実施（4点）

1. `memo_app.rb`で、`/memos/:id/edit`のURLに対して、404エラー処理を追加した。
2. `memo_app.rb`で、`not_found`メソッドで、機能が冗長的なところを削除した。
3. `memo.erb`で、フロントエンドのプログラムをシンプルにした（JavaScript外部ファイルの機能を、HTML内の`window.confirm()`に置き換えた）。
4. 不要な.jsファイルを削除した。

takao0206 commented:
修正したプログラムをpush致しました。
ご都合の良いときにご確認いただければ幸いです。

yoshitsugu commented on memo_app.rb:
```diff
@@ -0,0 +1,108 @@
+    erb :edit
+  else
+    status 404
+    erb :not_found
```
同様に、こちらも `erb` の記述をDRYにできると思います。

takao0206 replied:
私の理解が不十分で、申し訳ございません。
再度確認して、修正させていただきます。

yoshitsugu reviewed (commented):
コメントしました！

takao0206 committed:
404エラーに関するコードの重複部分を再修正した。

takao0206 commented:
ご指摘いただいたところを修正し、動作確認しました。
エラーのステータスとして「404」や「status 404」を指定すると、プログラム内の`not_found`が呼び出されることを理解致しました。
（自分で動作確認した後、すでにコメントに答えを書いて下さっていたことに気が付きました。sinatraの仕組みの理解が深まりました。ありがとうございました）

yoshitsugu reviewed (approved):
OKです！
